plutus-conformance/src/PlutusConformance/Common.hs

```diff
@@ -104,20 +39,147 @@ shownParseError = "parse error"
+                            (updateGoldenFile goldenFilePath) -- update the golden file
+            -- if the test is expected to fail, mark it so.
+            if expectedFailureFn dir
+            then pure $ expectFail mkGoldenTest
+            -- the test isn't expected to fail, make the `TestTree` as usual.
+            else pure mkGoldenTest
+    -- has children, so it's a grouping directory
+    else testGroup name <$> traverse (discoverTests eval expectedFailureFn) subdirs
+
+-- | Turn the expected file content in text to a `UplcProg` in Debruijn unless the expected result
+-- is a parse or evaluation error.
+expectedToProg :: T.Text -> Either T.Text UplcProg
+expectedToProg txt
+  | txt == shownEvaluationFailure =
+    Left txt
+  | txt == shownParseError =
+    Left txt
+  | otherwise =
+    case parseTxt txt of
+        Left _     -> Left txt
+        Right p -> -- The evaluator throws if the term has free variables
+            case UPLC.deBruijnTerm (UPLC._progTerm (void p)) of
+                Left (_ :: UPLC.FreeVariableError) -> Left shownEvaluationFailure
+                Right _                            -> Right $ void p
+
+-- | Get the tested value (in `Either T.Text UplcProg`).
+getTestedValue ::
+    UplcEvaluator
+    -> FilePath
+    -> IO (Either T.Text UplcProg)
+getTestedValue eval dir = do
+    inputFile <- findByExtension [".uplc"] dir
+    case inputFile of
+        [] -> error $ "Input file missing in " <> dir
+        _:_:_ -> error $ "More than 1 input files in " <> dir
+        [file] -> do
+            input <- T.readFile file
+            case parseTxt input of
+                Left _ -> pure $ Left shownParseError
+                Right p -> do
+                    case eval (void p) of
+                        Nothing                           -> pure $ Left shownEvaluationFailure
+                        Just prog@(UPLC.Program () _version tm) -> do
+                            case UPLC.deBruijnTerm tm of
+                                Left (_ :: UPLC.FreeVariableError) ->
+                                    pure $ Left shownEvaluationFailure
+                                Right _                         ->
+                                    pure $ Right prog
+
+-- | The comparison function used for the golden test.
+-- This function checks alpha-equivalence of programs when the output is a program.
+compareAlphaEq ::
+    Either T.Text UplcProg -- ^ golden value
+    -> Either T.Text UplcProg -- ^ tested value
+    -> Maybe String
+    -- ^ If two values are the same, it returns `Nothing`.
+    -- If they are different, it returns an error that will be printed to the user.
+compareAlphaEq (Left expectedTxt) (Left actualTxt) =
+    if actualTxt == expectedTxt
+    then Nothing
+    else Just $
+        "Test failed, the output failed to parse or evaluate: \n"
+        <> T.unpack actualTxt
+compareAlphaEq (Right expected) (Right actual) =
+    if actual == expected
```

Comment:
How does this work? You are supposed to check whether `actual` and `expected` are alpha-equivalent; why does `==` do the job? AFAICT, this does the same thing as before.

Reply:
> why does `==` do the job? 

The `Eq` instance for `Term Name uni fun ann` checks modulo alpha.

Reply:
:+1: